backends/dpdk/dpdkContext.cpp

```diff
@@ -0,0 +1,582 @@
+            add_space(out, 12);
+        }
+    }
+    out << "],\n";
+
+    // If table implementation is action profile or action selector, all actions from member
+    // table should be output for the base table.
+    const IR::P4Table *table = nullptr;
+    if (hasActionProfileSelector) {
+        table = memberTable;
+    } else {
+        table = tbl;
+    }
+
+    setActionAttributes(actionAttrMap, table);
+    cstring default_action_name = "";
+
+    unsigned default_action_handle = 0;
+    bool first = true;
+    if (table->getDefaultAction())
+        default_action_name = toStr(table->getDefaultAction());
+
+    // Print actions associated with this table
+    add_space(out, 12); out << "\"actions\": [";
+    for (auto action : table->getActionList()->actionList) {
+        struct actionAttributes attr = ::get(actionAttrMap, action->getName());
+        if (toStr(action->expression) == default_action_name)
+            default_action_handle = attr.actionHandle;
+
+        // Printing compiler added actions is curently not required
+        if (!attr.is_compiler_added_action) {
+            if (!first) out << ",";
+            out << "\n";
+            add_space(out, 16); out << "{\n";
+            auto actName = toStr(action->expression);
+
+            // NoAction is not prefixed with control block name
+            if (actName != "NoAction")
+                actName = mainTableAttr.controlName + "." + actName;
+            add_space(out, 20); out << "\"name\": \"" << actName << "\",\n";
+            add_space(out, 20); out << "\"handle\": " << attr.actionHandle << ",\n";
+            add_space(out, 20); out << "\"constant_default_action\": " <<
+                                       std::boolalpha << attr.constant_default_action << ",\n";
+            add_space(out, 20); out << "\"is_compiler_added_action\": " <<
+                                       std::boolalpha << attr.is_compiler_added_action  << ",\n";
+            add_space(out, 20); out << "\"allowed_as_hit_action\": " <<
+                                       std::boolalpha << attr.allowed_as_hit_action << ",\n";
+            add_space(out, 20); out << "\"allowed_as_default_action\": " <<
+                                       std::boolalpha << attr.allowed_as_default_action << ",\n";
+
+            // Print action parameters
+            add_space(out, 20); out << "\"p4_parameters\": [";
+            int position = 0;
+            if (attr.params) {
+                int index = 0;
+                for (auto param : *(attr.params)) {
+                    if (position) out << ",";
+                    out << "\n";
+                    add_space(out, 24); out << "{\n";
+                    add_space(out, 28); out << "\"name\": \""
+                                            << param->name.originalName << "\",\n";
+                    add_space(out, 28); out << "\"start_bit\": 0,\n";
+                    add_space(out, 28); out << "\"bit_width\": "
```

Comment:
are you checking someplace whether the parameters all have scalar types?

Reply:
there is a request to support structs for actions arguments.

Reply:
Added checks